File: ruby-pattern/TemplateMethod/TemplateMethod.rb

class TemplateMethod

  def initialize

  end

  def hoge
    hoge1
    print("\n")
    hoge2
    print("\n")
    hoge3
    print("\n")
  end

  def hoge1
  end

  def hoge2
  end

  def hoge3
  end

end

class TemplateMethodSub1 < TemplateMethod
  def hoge1
    print("--hoge1--")
  end

  def hoge2
    print("--hoge2--")
  end

  def hoge3
    print("--hoge3--")
  end

end

class TemplateMethodSub2 < TemplateMethod
  def hoge1
    print("**hoge1**")
  end

  def hoge2
    print("**hoge2**")
  end

  def hoge3
    print("**hoge3**")
  end

end

tm = TemplateMethod.new()
tm.hoge


tm1 = TemplateMethodSub1.new()
tm1.hoge

tm2 = TemplateMethodSub2.new()
tm2.hoge

File: swift-pattern/TemplateMethod.playground/Sources/TemplateMethod.swift

public class TemplateMethod{

    public init() {}
    
    public func hoge(){
        hoge1()
        hoge2()
        hoge3()
    }
    public func hoge1(){}
    public func hoge2(){}
    public func hoge3(){}

}

public class TemplateMethodSub1:TemplateMethod{
    public override func hoge1() {
        print("hoge1")
    }

    public override func hoge2() {
        print("hoge2")
    }

    public override func hoge3() {
        print("hoge3")
    }

}


public class TemplateMethodSub2:TemplateMethod{
    public override func hoge1() {
        print("*hoge1")
    }

    public override func hoge2() {
        print("*hoge2")
    }

    public override func hoge3() {
        print("*hoge3")
    }
    
}

